### PR TITLE
HPCC-22683 Add support for Restricted flag to Output and BuildIndex

### PR DESCRIPTION
**Pull request by shamser**
…Index

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday I am working on developing some regression programs for the new functionality.  Please can you start review?

File: ecl/hql/hqlattr.cpp [hunk 1]
```diff
@@ -631,7 +631,7 @@ unsigned getOperatorMetaFlags(node_operator op)
     case no_dataset_from_transform:
     case no_id:
 
-    case no_unused6:
+    case no_restricted:
```
Comment (ghalliday) [inline ecl/hql/hqlattr.cpp]: not used

File: ecl/hql/hqlexpr.cpp [hunk 2]
```diff
@@ -1943,7 +1943,7 @@ const char *getOpString(node_operator op)
     case no_orderedactionlist: return "ORDERED";
     case no_unordered: return "UNORDERED";
 
-    case no_unused6:
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: not used

File: ecl/hql/hqlexpr.hpp [hunk 3]
```diff
@@ -645,7 +645,7 @@ enum node_operator : unsigned short {
         no_cluster,
         no_forcenolocal,
         no_allnodes,
-        no_unused6,
+        no_restricted,
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.hpp]: not used

File: ecl/hql/reservedwords.cpp [hunk 4]
```diff
@@ -385,6 +385,7 @@ static const char * eclReserved12[] = {//Attributes
     "probability",
     "pulled",
     "remote",
+    "restricted",
```
Comment (ghalliday) [inline ecl/hql/reservedwords.cpp]: unused (if no_reserved is removed)

File: rtl/include/eclhelper.hpp [hunk 5]
```diff
@@ -1183,6 +1184,7 @@ enum
     TIWhaswidth         = 0x0800,
     TIWexpires          = 0x1000,
     TIWmaxlength        = 0x2000,       // explicit maxlength
+    TIWrestricted       = 0x200000,
```
Comment (ghalliday) [inline rtl/include/eclhelper.hpp]: Please add a comment to indicate this matches TDWrestricted.

File: ecl/hthor/hthor.cpp [hunk 6]
```diff
@@ -713,6 +713,7 @@ void CHThorDiskWriteActivity::publish()
         properties.setPropInt64("@totalCRC", totalCRC);
     }
     properties.setPropInt("@formatCrc", helper.getFormatCrc());
+    properties.setPropBool("restricted", (helper.getFlags() & TDWrestricted)!=0);
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: @jakesmith Should this always be set, or should it only set if restricted is true?

**Review by ghalliday — COMMENTED** [review]
@shamser looks sensible, 
- some old code needs removing.  
- Not sure if setProp("resticted", false) should be called.

File: ecl/hthor/hthor.cpp (hunk 6 above)
Comment (jakesmith) [reply ecl/hthor/hthor.cpp]: Yes, I would only bother setting the property if true.
So:
```
if (helper.getFlags() & TDWrestricted)
    properties.setPropBool("restricted", true);
```

**Pull request merged**